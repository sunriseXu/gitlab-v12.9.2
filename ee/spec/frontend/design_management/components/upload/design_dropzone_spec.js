import { shallowMount } from '@vue/test-utils';
import DesignDropzone from 'ee/design_management/components/upload/design_dropzone.vue';
import createFlash from '~/flash';

jest.mock('~/flash');

describe('Design management dropzone component', () => {
  let wrapper;

  const mockDragEvent = ({ types = ['Files'], files = [] }) => {
    return { dataTransfer: { types, files } };
  };

  const findDropzoneCard = () => wrapper.find('.design-dropzone-card');

  function createComponent({ slots = {}, data = {} } = {}) {
    wrapper = shallowMount(DesignDropzone, {
      slots,
      data() {
        return data;
      },
    });
  }

  afterEach(() => {
    wrapper.destroy();
  });

  describe('when slot provided', () => {
    it('renders dropzone with slot content', () => {
      createComponent({
        slots: {
          default: ['<div>dropzone slot</div>'],
        },
      });

      expect(wrapper.element).toMatchSnapshot();
    });
  });

  describe('when no slot provided', () => {
    it('renders default dropzone card', () => {
      createComponent();

      expect(wrapper.element).toMatchSnapshot();
    });

    it('triggers click event on file input element when clicked', () => {
      createComponent();
      const clickSpy = jest.spyOn(wrapper.find('input').element, 'click');

      findDropzoneCard().trigger('click');
      expect(clickSpy).toHaveBeenCalled();
    });
  });

  describe('when dragging', () => {
    it.each`
      description                  | eventPayload
      ${'is empty'}                | ${{}}
      ${'contains text'}           | ${mockDragEvent({ types: ['text'] })}
      ${'contains files and text'} | ${mockDragEvent({ types: ['Files', 'text'] })}
      ${'contains files'}          | ${mockDragEvent({ types: ['Files'] })}
    `('renders correct template when drag event $description', ({ eventPayload }) => {
      createComponent();

      wrapper.trigger('dragenter', eventPayload);
      return wrapper.vm.$nextTick().then(() => {
        expect(wrapper.element).toMatchSnapshot();
      });
    });

    it('renders correct template when dragging stops', () => {
      createComponent();

      wrapper.trigger('dragenter');
      return wrapper.vm
        .$nextTick()
        .then(() => {
          wrapper.trigger('dragleave');
          return wrapper.vm.$nextTick();
        })
        .then(() => {
          expect(wrapper.element).toMatchSnapshot();
        });
    });
  });

  describe('when dropping', () => {
    it('emits upload event', () => {
      createComponent();
      const mockFile = { name: 'test', type: 'image/jpg' };
      const mockEvent = mockDragEvent({ files: [mockFile] });

      wrapper.trigger('dragenter', mockEvent);
      return wrapper.vm
        .$nextTick()
        .then(() => {
          wrapper.trigger('drop', mockEvent);
          return wrapper.vm.$nextTick();
        })
        .then(() => {
          expect(wrapper.emitted().change[0]).toEqual([[mockFile]]);
        });
    });
  });

  describe('ondrop', () => {
    const mockData = { dragCounter: 1, isDragDataValid: true };

    describe('when drag data is valid', () => {
      it('emits upload event for valid files', () => {
        createComponent({ data: mockData });

        const mockFile = { type: 'image/jpg' };
        const mockEvent = mockDragEvent({ files: [mockFile] });

        wrapper.vm.ondrop(mockEvent);
        expect(wrapper.emitted().change[0]).toEqual([[mockFile]]);
      });

      it('calls createFlash when files are invalid', () => {
        createComponent({ data: mockData });

        const mockEvent = mockDragEvent({ files: [{ type: 'audio/midi' }] });

        wrapper.vm.ondrop(mockEvent);
        expect(createFlash).toHaveBeenCalledTimes(1);
      });
    });
  });
});
