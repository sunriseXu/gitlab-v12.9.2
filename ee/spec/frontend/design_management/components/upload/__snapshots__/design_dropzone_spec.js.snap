// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Design management dropzone component when dragging renders correct template when drag event contains files 1`] = `
<div
  class="w-100 position-relative"
>
  <button
    class="card design-dropzone-card design-dropzone-border w-100 h-100 d-flex-center p-3"
  >
    <div
      class="d-flex-center flex-column text-center"
    >
      <gl-icon-stub
        class="mb-4"
        name="doc-new"
        size="48"
      />
       
      <p>
        <gl-sprintf-stub
          message="%{lineOneStart}Drag and drop to upload your designs%{lineOneEnd} or %{linkStart}click to upload%{linkEnd}."
        />
      </p>
    </div>
  </button>
   
  <input
    accept="image/*"
    class="hide"
    multiple="multiple"
    name="design_file"
    type="file"
  />
   
  <transition-stub
    name="design-dropzone-fade"
  >
    <div
      class="card design-dropzone-border design-dropzone-overlay w-100 h-100 position-absolute d-flex-center p-3 bg-white"
      style=""
    >
      <div
        class="mw-50 text-center"
        style="display: none;"
      >
        <h3>
          Oh no!
        </h3>
         
        <span>
          You are trying to upload something other than an image. Please upload a .png, .jpg, .jpeg, .gif, .bmp, .tiff or .ico.
        </span>
      </div>
       
      <div
        class="mw-50 text-center"
        style=""
      >
        <h3>
          Incoming!
        </h3>
         
        <span>
          Drop your designs to start your upload.
        </span>
      </div>
    </div>
  </transition-stub>
</div>
`;

exports[`Design management dropzone component when dragging renders correct template when drag event contains files and text 1`] = `
<div
  class="w-100 position-relative"
>
  <button
    class="card design-dropzone-card design-dropzone-border w-100 h-100 d-flex-center p-3"
  >
    <div
      class="d-flex-center flex-column text-center"
    >
      <gl-icon-stub
        class="mb-4"
        name="doc-new"
        size="48"
      />
       
      <p>
        <gl-sprintf-stub
          message="%{lineOneStart}Drag and drop to upload your designs%{lineOneEnd} or %{linkStart}click to upload%{linkEnd}."
        />
      </p>
    </div>
  </button>
   
  <input
    accept="image/*"
    class="hide"
    multiple="multiple"
    name="design_file"
    type="file"
  />
   
  <transition-stub
    name="design-dropzone-fade"
  >
    <div
      class="card design-dropzone-border design-dropzone-overlay w-100 h-100 position-absolute d-flex-center p-3 bg-white"
      style=""
    >
      <div
        class="mw-50 text-center"
        style="display: none;"
      >
        <h3>
          Oh no!
        </h3>
         
        <span>
          You are trying to upload something other than an image. Please upload a .png, .jpg, .jpeg, .gif, .bmp, .tiff or .ico.
        </span>
      </div>
       
      <div
        class="mw-50 text-center"
        style=""
      >
        <h3>
          Incoming!
        </h3>
         
        <span>
          Drop your designs to start your upload.
        </span>
      </div>
    </div>
  </transition-stub>
</div>
`;

exports[`Design management dropzone component when dragging renders correct template when drag event contains text 1`] = `
<div
  class="w-100 position-relative"
>
  <button
    class="card design-dropzone-card design-dropzone-border w-100 h-100 d-flex-center p-3"
  >
    <div
      class="d-flex-center flex-column text-center"
    >
      <gl-icon-stub
        class="mb-4"
        name="doc-new"
        size="48"
      />
       
      <p>
        <gl-sprintf-stub
          message="%{lineOneStart}Drag and drop to upload your designs%{lineOneEnd} or %{linkStart}click to upload%{linkEnd}."
        />
      </p>
    </div>
  </button>
   
  <input
    accept="image/*"
    class="hide"
    multiple="multiple"
    name="design_file"
    type="file"
  />
   
  <transition-stub
    name="design-dropzone-fade"
  >
    <div
      class="card design-dropzone-border design-dropzone-overlay w-100 h-100 position-absolute d-flex-center p-3 bg-white"
      style=""
    >
      <div
        class="mw-50 text-center"
      >
        <h3>
          Oh no!
        </h3>
         
        <span>
          You are trying to upload something other than an image. Please upload a .png, .jpg, .jpeg, .gif, .bmp, .tiff or .ico.
        </span>
      </div>
       
      <div
        class="mw-50 text-center"
        style="display: none;"
      >
        <h3>
          Incoming!
        </h3>
         
        <span>
          Drop your designs to start your upload.
        </span>
      </div>
    </div>
  </transition-stub>
</div>
`;

exports[`Design management dropzone component when dragging renders correct template when drag event is empty 1`] = `
<div
  class="w-100 position-relative"
>
  <button
    class="card design-dropzone-card design-dropzone-border w-100 h-100 d-flex-center p-3"
  >
    <div
      class="d-flex-center flex-column text-center"
    >
      <gl-icon-stub
        class="mb-4"
        name="doc-new"
        size="48"
      />
       
      <p>
        <gl-sprintf-stub
          message="%{lineOneStart}Drag and drop to upload your designs%{lineOneEnd} or %{linkStart}click to upload%{linkEnd}."
        />
      </p>
    </div>
  </button>
   
  <input
    accept="image/*"
    class="hide"
    multiple="multiple"
    name="design_file"
    type="file"
  />
   
  <transition-stub
    name="design-dropzone-fade"
  >
    <div
      class="card design-dropzone-border design-dropzone-overlay w-100 h-100 position-absolute d-flex-center p-3 bg-white"
      style=""
    >
      <div
        class="mw-50 text-center"
      >
        <h3>
          Oh no!
        </h3>
         
        <span>
          You are trying to upload something other than an image. Please upload a .png, .jpg, .jpeg, .gif, .bmp, .tiff or .ico.
        </span>
      </div>
       
      <div
        class="mw-50 text-center"
        style="display: none;"
      >
        <h3>
          Incoming!
        </h3>
         
        <span>
          Drop your designs to start your upload.
        </span>
      </div>
    </div>
  </transition-stub>
</div>
`;

exports[`Design management dropzone component when dragging renders correct template when dragging stops 1`] = `
<div
  class="w-100 position-relative"
>
  <button
    class="card design-dropzone-card design-dropzone-border w-100 h-100 d-flex-center p-3"
  >
    <div
      class="d-flex-center flex-column text-center"
    >
      <gl-icon-stub
        class="mb-4"
        name="doc-new"
        size="48"
      />
       
      <p>
        <gl-sprintf-stub
          message="%{lineOneStart}Drag and drop to upload your designs%{lineOneEnd} or %{linkStart}click to upload%{linkEnd}."
        />
      </p>
    </div>
  </button>
   
  <input
    accept="image/*"
    class="hide"
    multiple="multiple"
    name="design_file"
    type="file"
  />
   
  <transition-stub
    name="design-dropzone-fade"
  >
    <div
      class="card design-dropzone-border design-dropzone-overlay w-100 h-100 position-absolute d-flex-center p-3 bg-white"
      style="display: none;"
    >
      <div
        class="mw-50 text-center"
      >
        <h3>
          Oh no!
        </h3>
         
        <span>
          You are trying to upload something other than an image. Please upload a .png, .jpg, .jpeg, .gif, .bmp, .tiff or .ico.
        </span>
      </div>
       
      <div
        class="mw-50 text-center"
        style="display: none;"
      >
        <h3>
          Incoming!
        </h3>
         
        <span>
          Drop your designs to start your upload.
        </span>
      </div>
    </div>
  </transition-stub>
</div>
`;

exports[`Design management dropzone component when no slot provided renders default dropzone card 1`] = `
<div
  class="w-100 position-relative"
>
  <button
    class="card design-dropzone-card design-dropzone-border w-100 h-100 d-flex-center p-3"
  >
    <div
      class="d-flex-center flex-column text-center"
    >
      <gl-icon-stub
        class="mb-4"
        name="doc-new"
        size="48"
      />
       
      <p>
        <gl-sprintf-stub
          message="%{lineOneStart}Drag and drop to upload your designs%{lineOneEnd} or %{linkStart}click to upload%{linkEnd}."
        />
      </p>
    </div>
  </button>
   
  <input
    accept="image/*"
    class="hide"
    multiple="multiple"
    name="design_file"
    type="file"
  />
   
  <transition-stub
    name="design-dropzone-fade"
  >
    <div
      class="card design-dropzone-border design-dropzone-overlay w-100 h-100 position-absolute d-flex-center p-3 bg-white"
      style="display: none;"
    >
      <div
        class="mw-50 text-center"
      >
        <h3>
          Oh no!
        </h3>
         
        <span>
          You are trying to upload something other than an image. Please upload a .png, .jpg, .jpeg, .gif, .bmp, .tiff or .ico.
        </span>
      </div>
       
      <div
        class="mw-50 text-center"
        style="display: none;"
      >
        <h3>
          Incoming!
        </h3>
         
        <span>
          Drop your designs to start your upload.
        </span>
      </div>
    </div>
  </transition-stub>
</div>
`;

exports[`Design management dropzone component when slot provided renders dropzone with slot content 1`] = `
<div
  class="w-100 position-relative"
>
  <div>
    dropzone slot
  </div>
   
  <transition-stub
    name="design-dropzone-fade"
  >
    <div
      class="card design-dropzone-border design-dropzone-overlay w-100 h-100 position-absolute d-flex-center p-3 bg-white"
      style="display: none;"
    >
      <div
        class="mw-50 text-center"
      >
        <h3>
          Oh no!
        </h3>
         
        <span>
          You are trying to upload something other than an image. Please upload a .png, .jpg, .jpeg, .gif, .bmp, .tiff or .ico.
        </span>
      </div>
       
      <div
        class="mw-50 text-center"
        style="display: none;"
      >
        <h3>
          Incoming!
        </h3>
         
        <span>
          Drop your designs to start your upload.
        </span>
      </div>
    </div>
  </transition-stub>
</div>
`;
